pepr/utilities.py

```diff
@@ -51,3 +55,282 @@ def assign_record_ids_to_target_models(
             )
 
     return records_per_target_model + offset
+
+
+def filter_out_outlier(data, labels, filter_pars, save_path=None, load_pars=None):
+    """
+    Filter out potentially vulnerable samples.
+
+    The general idea is similar to (:meth:`pepr.privacy.gmia.DirectGmia`) to find
+    potential vulnerable records.
+
+    Steps:
+
+    1. Create mapping of records to reference models.
+    2. Train the reference models.
+    3. Generate intermediate models.
+    4. Extract high-level features.
+    5. Compute pairwise distances between high-level features.
+    6. Determine potential vulnerable records.
+    7. Remove potential vulnerable records.
+
+    Parameters
+    ----------
+    data : numpy.ndarray
+        Dataset with all training samples used in the given pentesting setting.
+    labels : numpy.ndarray
+        Array of all labels used in the given pentesting setting.\
+    filter_pars : dict
+        Dictionary containing needed filter parameters:
+
+        * number_classes (int): Number of different classes the target model
+          predict.
+        * number_reference_models (int): Number of reference models to be trained.
+        * reference_training_set_size (int): Size of the trainings set for each
+          reference model.
+        * create_compile_model (function): Function that returns a compiled
+          TensorFlow model (typically identical to the target model) used in the
+          training of the reference models.
+        * reference_epochs (int): Number of training epochs of the reference models.
+        * reference_batch_size (int): Batch size used in the training of the
+          reference models.
+        * hlf_metric (str): Metric (typically 'cosine') used for the distance
+          calculations in the high-level feature space. For valid metrics see
+          documentation of sklearn.metrics.pairwise_distances.
+        * hlf_layer_number (int): If value is n, the n-th layer of the model
+          returned by create_compile_model is used to extract the high-level feature
+          vectors.
+        * distance_neighbor_threshold (float): If distance is smaller than the neighbor
+          threshold the record is selected as target record.
+        * number_neighbor_threshold (float): If number of neighbors of a record is
+          smaller than this, it is considered a vulnerable example.
+        * number_outlier (int): If set, the selection algorithm performs
+          `max_search_rounds`, to find a `distance_neighbor_threshold`, that leads to a
+          finding of `n_targets` target records. These target records are most
+          vulnerable with respect to our selection criterion.
```

Comment:
I think if you like you can move your nice description of the initialization of the ``neighbor_threshold``to this bullet point 👍

Reply:
I think it's an implementation detail that does not need to be in the documentation